### DISCUSSION TITLE
Help With Codes in Message-IX

### DISCUSSION THREAD

**Discussion by Anurag-Dey** [iiasa/message_ix#481] [MESSAGE modeling]
Hi All,

I am building a simple power sector model for India based on the Westeros model in Message-IX but modified according to the requirement. I have already input various variables such as investment cost, fixed O&M and many more. Now I need help with the following - 
a) The Westeros Model shows Natural Resources in the Primary Energy Level. I need help with the codes to divide the Natural Resources into coal domestic production and Coal Import in the Primary Energy Level. And then adding the fuel availability and fuel cost data for coal in the model. 
b) The Westeros Model does not show Primary Level while adding all the Energy Levels to the scenario, although the Reference Energy System (RES) shows natural resources as the Primary Energy. I need clarity in this regard. Also, I need help with the codes to add a Primary Energy Level in the Scenario.
c) I need clarity on the codes to add base_activity and bounds in the Austrian Model. I have attached the snapshot below. I want to know what the technology wise value suggests. Is it data for the fixed base_activity or is it the data for upper or lower bound ? If it is the fixed activity data, then how can I add upper or lower bound to the data and if it is bound data, how can I add the fixed activity data ? (Kindly Explain for Time Series as well).
 
![image](https://user-images.githubusercontent.com/66878460/120597328-86a91880-c462-11eb-8804-6f168d9d97d6.png)

d) Do you have a sample extensive model which shows all the possible codes for all the possible variables in the Power Sector (any country) ? Just like Westeros and Austrian Model but cover a larger set of variables.

Thanks & Regards,
Anurag Dey

**Comment by khaeru** [comment]
Some tips when seeking help:
- Please use [GitHub's Markdown formatting](https://guides.github.com/features/mastering-markdown/), in particular code fencing with triple backticks (\`\`\`, described [here](https://guides.github.com/features/mastering-markdown/#examples))—do not insert screenshots, as these are not accessible.
- Since this is neither about (a) a bug or error in the MESSAGEix package nor (b) a requested feature in the software per se, I have converted this from an issue to a “Discussion”. As the documentation indicates, please only open issues if (a) or (b) is the case; otherwise use the Discussions feature.

**Comment by OFR-IIASA** [comment]
In response to point "_a) The Westeros Model shows Natural Resources in the Primary Energy Level. I need help with the codes to divide the Natural Resources into coal domestic production and Coal Import in the Primary Energy Level. And then adding the fuel availability and fuel cost data for coal in the model._"

There is a tutorial available for adding fossil resources (westeros\westeros_fossil_resource.ipynb). You will also find the documentation of resource related equations [here](https://docs.messageix.org/en/stable/model/MESSAGE/model_core.html#resource-and-commodity-section).

The tutorial covers the main parameters associated with fossil resources:
- Adding varying types of coal resource types for example via the parameter `grade`.
- Defining the available volumes of resources for the individual grades via the parameter `resource_potential`.
- Adding resource costs via the parameter `resource_cost`.
- Adding resource extraction that has already taken place in the past `historical_extraction`.

The import of coal can depicted as a technology as opposed to being depicted as a `resource`. So, using activity bounds and variable costs to configure the import quantities available and the therewith associated costs.

Step "_1. Defining level, grade, and commodity_" tutorial will also explain how to address your second question "_b) The Westeros Model does not show Primary Level while adding all the Energy Levels to the scenario, although the Reference Energy System (RES) shows natural resources as the Primary Energy. I need clarity in this regard. Also, I need help with the codes to add a Primary Energy Level in the Scenario._". Therefore, a level can be added using the command `scen.add_set('level', <level_name>)`

As for the question relating to the Austria tutorial "_c) I need clarity on the codes to add base_activity and bounds in the Austrian Model. I have attached the snapshot below. I want to know what the technology wise value suggests. Is it data for the fixed base_activity or is it the data for upper or lower bound ? If it is the fixed activity data, then how can I add upper or lower bound to the data and if it is bound data, how can I add the fixed activity data ? (Kindly Explain for Time Series as well)._". As you can see at the bottom of the screenshot, the values, which represent the activity levels of individual technologies in the year 2010, are added as both `bound_activity_up` and `bound_activity_lo`, which is a way of fixing the activity. In order to extrapolate these values for additional years, you can modify the dataframe so that there are additional `year_act` and the corresponding `value` added for the respective years. `base_activity` is not a parameter in MESSAGEix, it is purely the name used to describe the variable/dataframe that we are creating in the example above. 

As for the last question "_d) Do you have a sample extensive model which shows all the possible codes for all the possible variables in the Power Sector (any country) ? Just like Westeros and Austrian Model but cover a larger set of variables._", we do not have an extensive model which is publicly available which specifically covers the power sector for a country. There is an example of a national model for South Africa available [here](https://github.com/tum-ewk/message_ix_south_africa), but **please keep in mind that this is not a repository which IIASA maintains, hence there is no guarantee that the provided code will work out of the box.**
